### ISSUE TITLE
Correct Rationale explanation for 0013 RLS disabled in public

### ISSUE THREAD

**Issue by autopaybyphone** [supabase/splinter#82]
# Improve documentation

## Link 

https://supabase.com/docs/guides/database/database-advisors?lint=0013_rls_disabled_in_public

## Describe the problem

The doc says 

> If row level security (RLS) is not enabled on a public table, anyone with the project's URL can CREATE/READ/UPDATE/DELETE (CRUD) rows in the impacted table

If our understanding is correct, that is only true if one has shared the API key publicly.

In our case, we don't use the API, we only connect to the database from our server using Supavisor's database string. No one is connecting to the database from their browser, the API key is not shared publicly.

Thus it seems that it is not true in our case that "anyone with the project's URL can CRUD rows".

We understand that the linter cannot know whether we have shared the API key publicly or not, but it seems to us that the doc should be more precise, so that people don't get confused.

## Describe the improvement

You could add a precision in a colored frame, that mentions that it is not the case if the API key is never shared, but then one should be sure of this. This precision could mention that in that case `public` can be removed from the "exposed schemas" settings of the API, to make the error disappear.

## Additional context

https://github.com/orgs/supabase/discussions/26584

**Comment by autopaybyphone** [comment]
Has anybody seen this?

**Comment by olirice** [comment]
You're right that an apiKey is needed to access your tables but that value is not sufficiently secure to be treated as a secret/password. If its ever used, it'd be readable in logs as a plaintext header which makes it inappropriate in a security context

> could mention that in that case public can be removed from the "exposed schemas"

thank you, good feedback. Adding in #83

**Comment by autopaybyphone** [comment]
Amazing, thanks for the very swift reaction!

Just to make sure we understand correctly: do you confirm that if we are connecting to the database only through the Supavisor URL from our server, then removing `public` from the exposed schemas won't break anything?

**Comment by olirice** [comment]
correct, if you aren't using REST/GraphQL it won't impact you

**Comment by autopaybyphone** [comment]
Awesome, thanks a lot!

**Comment by autopaybyphone** [comment]
We wanted to test removing `public` from exposed schemas, on our staging environment first. However, when logging in to our staging account, it seems that although the settings are exactly the same, the linter doesn't raise any error. There is no Row Level Security on any table either, `public` is in the exposed schemas settings of the API, and we clicked on "Rerun linter", still nothing.

Do you have any idea why that would be?

**Comment by olirice** [comment]
A quick way to run down what the difference is would be to take that lint's SQL

```sql
set local pgrst.db_schemas = 'public,graphql_public'; -- Adjust this to match your "Exposed Schemas" API setting

select
    'rls_disabled_in_public' as name,
    'RLS Disabled in Public' as title,
    'ERROR' as level,
    'EXTERNAL' as facing,
    array['SECURITY'] as categories,
    'Detects cases where row level security (RLS) has not been enabled on tables in schemas exposed to PostgREST' as description,
    format(
        'Table \`%s.%s\` is public, but RLS has not been enabled.',
        n.nspname,
        c.relname
    ) as detail,
    'https://supabase.com/docs/guides/database/database-linter?lint=0013_rls_disabled_in_public' as remediation,
    jsonb_build_object(
        'schema', n.nspname,
        'name', c.relname,
        'type', 'table'
    ) as metadata,
    format(
        'rls_disabled_in_public_%s_%s',
        n.nspname,
        c.relname
    ) as cache_key
from
    pg_catalog.pg_class c
    join pg_catalog.pg_namespace n
        on c.relnamespace = n.oid
where
    c.relkind = 'r' -- regular tables
    -- RLS is disabled
    and not c.relrowsecurity
    and (
        pg_catalog.has_table_privilege('anon', c.oid, 'SELECT')
        or pg_catalog.has_table_privilege('authenticated', c.oid, 'SELECT')
    )
    and n.nspname = any(array(select trim(unnest(string_to_array(current_setting('pgrst.db_schemas', 't'), ',')))))
    and n.nspname not in (
        '_timescaledb_cache', '_timescaledb_catalog', '_timescaledb_config', '_timescaledb_internal', 'auth', 'cron', 'extensions', 'graphql', 'graphql_public', 'information_schema', 'net', 'pgroonga', 'pgsodium', 'pgsodium_masks', 'pgtle', 'pgbouncer', 'pg_catalog', 'pgtle', 'realtime', 'repack', 'storage', 'supabase_functions', 'supabase_migrations', 'tiger', 'topology', 'vault'
    );
```
and drop it into Studio's SQL editor

When you run it in your prod instance, you'll see the records. When you run it in staging, they should be missing (to reproduce the behavior you're currently seeing).

Then you can comment out the `where` clauses 1 at a time in your staging environment and see which one causes the missing record to show up.

**Comment by autopaybyphone** [comment]
Awesome, thank a lot! We'll try that!

**Comment by autopaybyphone** [comment]
So it turns out it's this part that removes the `public` tables in our staging area:

```
and (
        pg_catalog.has_table_privilege('anon', c.oid, 'SELECT')
        or pg_catalog.has_table_privilege('authenticated', c.oid, 'SELECT')
    )
```

However, we don't remember ever messing with the privileges, neither in our production nor in our staging database. Do you have any idea how that could have diverged?

Thanks a lot again for the help!

**Comment by olirice** [comment]
Unfortunately I couldn't say why the permissions might have diverged.

The default permissions config gives select access to `anon` and `authenticated` in the `public` schema.
It sounds like that might have been revoked on your staging instance at some point.

If you can easily rebuild your staging instance it might be worth moving it to a new project to true-up the permissions

**Comment by olirice** [comment]
If you're using prisma that can often interact with permissions unless you followed the [guide](https://supabase.com/partners/integrations/prisma)

**Comment by autopaybyphone** [comment]
Thanks a lot for your help once again! We are not using Prisma, we will investigate. Thank you!

**Comment by autopaybyphone** [comment]
It's really weird. We started a new project on our staging environment, rebuilt the database just applying our migrations, and still no error.

Is it the default that "anon" and "authenticated" don't have privilege?

**Comment by olirice** [comment]
by default `anon` and `authenticated` have select access to your `public` schema. They do not have access to any other schemas by default

**Comment by autopaybyphone** [comment]
Thanks for the reply!

We'll open another issue then, because what we are experiencing seems contradictory.

Thanks again very much for your help!
